eth/watchers/senderwatcher_test.go

```diff
@@ -0,0 +1,571 @@
+	assert := assert.New(t)
+	stubClientSenderInfo := &pm.SenderInfo{
+		Deposit: big.NewInt(10),
+		Reserve: big.NewInt(5),
+	}
+	sw := &SenderWatcher{
+		senders: make(map[ethcommon.Address]*pm.SenderInfo),
+		lpEth: &eth.StubClient{
+			SenderInfo: stubClientSenderInfo,
+		},
+	}
+	info := &pm.SenderInfo{
+		Deposit:      big.NewInt(1000000000000000000),
+		Reserve:      big.NewInt(1000000000000000000),
+		ReserveState: pm.NotFrozen,
+	}
+	sender := pm.RandAddress()
+
+	sw.setSenderInfo(sender, info)
+
+	getInfo, err := sw.GetSenderInfo(sender)
+	assert.Nil(err)
+	assert.Equal(info, getInfo)
+
+	sw.Clear(sender)
+	time.Sleep(2 * time.Millisecond)
+
+	getInfo = sw.senders[sender]
+	assert.Nil(getInfo)
+	// on calling GetSenderInfo should be back to RPC stub values
+	getInfo, err = sw.GetSenderInfo(sender)
+	assert.Nil(err)
+	assert.Equal(getInfo, stubClientSenderInfo)
+}
+
+func TestSenderWatcher_WatchAndStop(t *testing.T) {
+	assert := assert.New(t)
+	lpEth := &eth.StubClient{}
+	watcher := &stubBlockWatcher{}
+	sw, err := NewSenderWatcher(stubTicketBrokerAddr, watcher, lpEth)
+	assert.Nil(err)
+
+	go sw.Watch()
+	time.Sleep(2 * time.Millisecond)
+	sw.Stop()
+	time.Sleep(2 * time.Millisecond)
+	assert.True(watcher.sub.unsubscribed)
+}
+
+func TestSenderWatcher_HandleLog(t *testing.T) {
+	lpEth := &eth.StubClient{}
+	watcher := &stubBlockWatcher{}
+	rw, err := NewSenderWatcher(stubTicketBrokerAddr, watcher, lpEth)
+	require.Nil(t, err)
+
+	assert := assert.New(t)
+
+	// Test unknown event
+	log := newStubBaseLog()
+	log.Topics = []ethcommon.Hash{ethcommon.BytesToHash([]byte("foo"))}
+
+	err = rw.handleLog(log)
+	assert.Nil(err)
+}
+
+func TestFundDepositEvent(t *testing.T) {
+	assert := assert.New(t)
+	require := require.New(t)
+	senderInfo := &pm.SenderInfo{
+		Deposit: big.NewInt(10),
+		Reserve: big.NewInt(5),
+	}
+	lpEth := &eth.StubClient{
+		SenderInfo: senderInfo,
+	}
+	watcher := &stubBlockWatcher{}
+	sw, err := NewSenderWatcher(stubTicketBrokerAddr, watcher, lpEth)
+	assert.Nil(err)
+
+	header := defaultMiniHeader()
+	newDepositEvent := newStubDepositFundedLog()
+	header.Logs = append(header.Logs, newDepositEvent)
+
+	blockEvent := &blockwatch.Event{
+		Type:        blockwatch.Added,
+		BlockHeader: header,
+	}
+
+	go sw.Watch()
+	defer sw.Stop()
+	time.Sleep(2 * time.Millisecond)
+
+	// If map entry is nil don't handle event
+	watcher.sink <- []*blockwatch.Event{blockEvent}
+	time.Sleep(2 * time.Millisecond)
+	_, ok := sw.senders[stubSender]
+	assert.False(ok)
+
+	// If map entry exists parse event log
+	info, err := sw.GetSenderInfo(stubSender)
+	require.Nil(err)
+	require.Equal(info, senderInfo)
+	watcher.sink <- []*blockwatch.Event{blockEvent}
+	time.Sleep(2 * time.Millisecond)
+	expectedDeposit := senderInfo.Deposit.Add(senderInfo.Deposit, new(big.Int).SetBytes(newDepositEvent.Data))
+	info, err = sw.GetSenderInfo(stubSender)
+	assert.Nil(err)
+	assert.Zero(info.Deposit.Cmp(expectedDeposit))
```

Comment:
While this test is supposed to check that a sender's deposit increases by the amount indicated by the `DepositFunded` event, in reality it is checking that a sender's deposit increased by the amount indicated by the `DepositFunded` event multiplied by 2.

See below for the test output when we add print statements in the following locations:
- Print `senderInfo.Deposit` right before L156
- Print `senderInfo.Deposit` right before L158
- Print `expectedDeposit` right after L159
- Print `info.Deposit` and `expectedDeposit` right after L161

```
=== RUN   TestFundDepositEvent
senderInfo.Deposit before sending to sink: 10
senderInfo.Deposit after sending to sink: 5000000000000000010
expected deposit: 10000000000000000010
asserted that info.Deposit = expectedDeposit 10000000000000000010 = 10000000000000000010
```

Notice how `senderInfo.Deposit` increases by the `DepositFunded` amount after the block event is sent into the sink. But, then on L158, we mutate `senderInfo.Deposit` when calculating `expectedDeposit` such that it `senderInfo.Deposit` ends up equaling 10 + (the `DepositFunded` amount * 2). The assertion on L161 still passes because any updates to `senderInfo.Deposit` are reflected in the `info` variable returned on L159 since both `info` and `senderInfo` reference the same pointer.

Reply:
fixed in 6e29515